### PR TITLE
[Concurrency] Build executors as their own separate object modules.

### PR DESCRIPTION
**Pull request by al45tair**
We want to split the global executor implementations out into their own separate object files, because that will allow us to provide a custom executor specific to the target platform when using Embedded Concurrency.

rdar://135380149

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
Hmmm. Looks like I've broken a few tests there :-) More to do, it seems.

File: stdlib/public/Concurrency/CMakeLists.txt [hunk 1]
```diff
@@ -78,6 +74,7 @@ set(SWIFT_RUNTIME_CONCURRENCY_C_SOURCES
   AsyncLet.cpp
   Clock.cpp
   GlobalExecutor.cpp
+  Hooks.cpp
```
Comment (ktoso) [inline stdlib/public/Concurrency/CMakeLists.txt:77]: nitpick, maybe call the file ConcurrencyHooks.cpp?

**Review by ktoso — APPROVED** [review]
AFAICS this looks pretty right. The new header looks good and has all the hooks I'd expect.

File: stdlib/public/Concurrency/CMakeLists.txt (hunk 1 above)
Comment (al45tair) [reply stdlib/public/Concurrency/CMakeLists.txt]: Done.

**Review by rjmccall — COMMENTED** [review]
To what extent is this protocol for async main (the drain / donate system) already ABI vs. just an internal detail of the concurrency library?  That's the main thing that stands out to me — that we're effectively making this pseudo-ABI (admittedly still something we reserve the right to change in later releases), and we should consider what clients are likely to actually want to take over.

**Comment by al45tair** [comment]
@rjmccall That's an interesting point. I think `swift_task_asyncMainDrainQueue()` is _actually_ ABI, in that a call to it is emitted by the compiler when it builds the `async_Main` entry point in `SILGenFunction::emitAsyncMainThreadStart()`. I'm less sure about `swift_task_donateThreadToGlobalExecutorUntil()`, in that I don't see that being emitted by the compiler (but that might just be me failing to find it)… I assume that was added for some embedded use-case or other?

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
> @rjmccall That's an interesting point. I think `swift_task_asyncMainDrainQueue()` is _actually_ ABI, in that a call to it is emitted by the compiler when it builds the `async_Main` entry point in `SILGenFunction::emitAsyncMainThreadStart()`. I'm less sure about `swift_task_donateThreadToGlobalExecutorUntil()`, in that I don't see that being emitted by the compiler (but that might just be me failing to find it)… I assume that was added for some embedded use-case or other?

Yeah, it's otherwise impossible to actually drain the queue.

Okay, if `asyncMainDrainQueue` is called directly by the compiler, then it's what we've got to live with.

**Comment by al45tair** [comment]
Rebased.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
Okay.  Ideally I think that entry point would purely be an additional function offered when we build in that mode. And just in general I don't think we care about supporting any of the patching mechanisms in the embedded mode that uses this — they're really only there for the stable-ABI platforms.

**Pull request merged**